File: token/token.go
package token

// token identifier
type TokenKind string

type Token struct {
	Kind    TokenKind // token identifier
	Literal string    // data associated with the token
}

const (
	// critical control tokens
	Illegal = "Illegal" // signifies an unidentifiable token
	Eof     = "Eof"     // end of file or eof symbol //\n

	// identifiers and literals
	Identifier = "Identifier" // variable names
	Integer    = "Integer"    // int type
	String     = "String"     // string type

	// operators
	Bind = "=" // assignment operator
	Plus = "+" // addition and concatenation

	// delimiters
	Comma = "," // delimiting options

	// precedence tokens
	LParen = "("
	RParen = ")"

	// line type designators -> [//, //*, /*]
	Slash = "/"
	Aster = "*"

	// keywords
	Set  = "Set"
	Proc = "Proc"
	Exec = "Exec"
	Dd   = "Dd"
)

var keywords = map[string]TokenKind{
	"SET":  Set,
	"PROC": Proc,
	"EXEC": Exec,
	"DD":   Dd,
}

func LookupIdentifier(ident string) TokenKind {
	if tok, ok := keywords[ident]; ok {
		return tok
	}
	return Identifier
}

File: src/main/ast/Token.java
package main.ast;

import java.util.HashMap;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Token {

    public TokenKind kind;
    public String literal;

    public Token(TokenKind kind, String literal) {
        this.kind = kind;
        this.literal = literal;
    }

    private static final HashMap<String, TokenKind> Keywords = (HashMap<String, TokenKind>) Stream.of(new Object[][]{
                    {"SET",TokenKind.Set},
                    {"PROC",TokenKind.Proc},
                    {"EXEC",TokenKind.Exec},
                    {"DD",TokenKind.Dd},
            })
            .collect(Collectors.toMap(data -> (String) data[0], data -> (TokenKind)data[1]));

    public static Optional<TokenKind> lookupIdentifier(String identifier) {
        return Optional.ofNullable(Keywords.get(identifier));
    }

    @Override public String toString() {
        return this.kind.name() + ": " + this.literal;
    }
}




